### ISSUE TITLE
GoRouter not popping on Android 14 with android back button

### ISSUE THREAD

**Issue by theREMASTERED1** [flutter/flutter#145198]
### What package does this bug report belong to?

go_router

### What target platforms are you seeing this bug on?

Android

### Have you already upgraded your packages?

Yes

### Dependency versions

<details><summary>pubspec.lock</summary>

```lock
  go_router:
    dependency: "direct main"
    description:
      name: go_router
      sha256: "170c46e237d6eb0e6e9f0e8b3f56101e14fb64f787016e42edd74c39cf8b176a"
      url: "https://pub.dev"
    source: hosted
    version: "13.2.0"
```

</details>

### Steps to reproduce

1. add go_router to project,even earlier versions since v10
2. Make sure the device is running Android 14 (Tested versions of Android using Emulators and Physical devices)
3. create a routing system using go_Router
4. set the RouterConfig as the router that you created
5. try to press the back button on the Android and it will close or minimize the app instead of poping
6. try the same thing on Android 13 or earlier, Works perfectly

### Expected results

Pop To previous page and pressing of Back Button as on other android versions

### Actual results

Closes app or minimizes it

### Code sample

<details open><summary>Code sample</summary>

```dart
GoRouter router = GoRouter(
  initialLocation: LoginScreen.routeName,
  debugLogDiagnostics: true,
  routes: <RouteBase>[
    GoRoute(
      path: LoginScreen.routeName,
      name: LoginScreen.title,
      pageBuilder: (context, state) => MaterialPage<void>(
        key: state.pageKey,
        child: const LoginScreen(),
      ),
    ),
    GoRoute(
      path: PinScreen.routeName,
      name: PinScreen.title,
      pageBuilder: (context, state) => MaterialPage<void>(
        key: state.pageKey,
        child: const PinScreen(),
      ),
    ),
```

```
 return MaterialApp.router(
      scaffoldMessengerKey: scaffoldMessengerKey,
      title: 'My APP',
      localizationsDelegates: const [
        AppLocalizationsDelegate(),
        GlobalMaterialLocalizations.delegate,
        GlobalWidgetsLocalizations.delegate,
        GlobalCupertinoLocalizations.delegate,
      ],
      locale: AppLocalizations.getStoredLocale(),
      supportedLocales: const [
        Locale('en'),
        Locale('af'),
      ],
      theme: ThemeData(brightness: Brightness.light, useMaterial3: false),
      darkTheme: ThemeData(brightness: Brightness.dark, useMaterial3: false),
      routerConfig: router,
    );

```
    
Id like to add that I'm using PopScope but the router doesn't use the PopScope at all it seems

</details>

### Screenshots or Videos

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>
Usually the logs would be something like

```console
[GoRouter] popping to /home
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.16.4, on macOS 14.0 23A344 darwin-arm64, locale en-ZA)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] VS Code (version 1.87.1)
[✓] Connected device (2 available)
[✓] Network resources
```

</details>

**Comment by theREMASTERED1** [comment]
I'm Thining the GoRouter's backButtonDispatcher is not working well with Android 14

**Comment by huycozy** [comment]
Hi @theREMASTERED1, I think you encountered https://github.com/flutter/flutter/issues/138737. You can check the 1st comment to see if your case also is the same one of them. If your case is not the same, please provide a complete sample code that implemented PopScope. Thanks!

**Comment by theREMASTERED1** [comment]
Hi @huycozy I checked out the issue and updated the router's version to 12.1.3 . When I tested, I noticed something: the PopScope works fine, but the onPopInvoked only kicks in when canPop is false, especially on older Android versions.

Then I tried it on an Android 14 device with canPop set to false, but it didn't work as expected. Instead of not popping at all, like it should with the PopScope, the app just closed. It seemed like it was ignoring the PopScope altogether.

So it it seems my issue is kind of related but also not at all

here's a example of my PopScope
```
  @override
  Widget build(BuildContext context) {
    ref.read(snackBarProvider.notifier).hideCurrentSnackBar();
    return PopScope(
      canPop: true,
      onPopInvoked: (didPop) {
        print('didPop: $didPop');
      },
      ```

**Comment by huycozy** [comment]
Thanks for the update. I see `canPop` set to `true` in your example but you said `canPop` set to `false`. If `canPop` is true, it's indeed a case reported at https://github.com/flutter/flutter/issues/138737#issue-2002700595 (the reporter also tests on Android 14).

This may also related to https://github.com/flutter/flutter/issues/140869.

**Comment by theREMASTERED1** [comment]
Hi @huycozy.Yes I am experiencing **almost** the same issue, i turned canPop on and off to test it. the thing is what I'm saying is that Android 14 is ignoring the PopScope entirely regardless if its set to false or true, Its like Android 14 ignores popping on goRouter entirely

I'm having the issue you previously mentioned on the previous Android version but it is okay as don't use PopInvoked really
Even changing the Gorouter version is not really making a difference for me

so short version is that GoRouter hardware pop functionality seems to be ignored entirely on Android 14 regardless of the package version or if canPop is true or false
its just closes/mimimizes the app
(it does pop when i press in-app backbuttons)

**Comment by huycozy** [comment]
I'm more inclined to centralize on https://github.com/flutter/flutter/issues/138737 because it seems to encapsulate this as well. If you also agree, feel free to close this and follow that issue.

Otherwise, please make this more clear by specific test cases:
I did a test on my `Pixel 7 Android 14` device, `onPopInvoked` is always invoked regardless `canPop` is true or false. This does not match a test case at https://github.com/flutter/flutter/issues/138737#issue-2002700595 (when canPop is true). Could you also try with the [given example there](https://github.com/btrautmann/pop_scope_repro) and share the result in the table here?
